codec-http2/src/main/java/io/netty/handler/codec/http2/ReadOnlyHttp2Headers.java

```diff
@@ -0,0 +1,843 @@
+                                                     boolean validateHeaders) {
+        return new ReadOnlyHttp2Headers(new AsciiString[] { PseudoHeaderName.STATUS.value(), status }, otherHeaders,
+                                        validateHeaders);
+    }
+
+    private ReadOnlyHttp2Headers(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders, boolean validateHeaders) {
+        if ((pseudoHeaders.length & 1) != 0 || (otherHeaders.length & 1) != 0) {
+            throw newIllegalArgumentException();
+        }
+        if (validateHeaders) {
+            validateHeaders(pseudoHeaders, otherHeaders);
+        }
+        this.pseudoHeaders = pseudoHeaders;
+        this.otherHeaders = otherHeaders;
+    }
+
+    private static IllegalArgumentException newIllegalArgumentException() {
+        return new IllegalArgumentException("pseudoHeaders and otherHeaders must be arrays of [name, value] pairs");
+    }
+
+    private static void validateHeaders(AsciiString[] pseudoHeaders, AsciiString[] otherHeaders) {
+        boolean seenNonPseudoHeader = false;
+        int i;
+        AsciiString name;
+        for (i = 0; i < pseudoHeaders.length; ++i) {
+            name = pseudoHeaders[i];
+            HTTP2_NAME_VALIDATOR.validateName(name);
+            if (!seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) != ':') {
+                seenNonPseudoHeader = true;
+            } else if (seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) == ':') {
+                throw new IllegalArgumentException(
+                     "pseudoHeaders name at index " + i + " is a pseudo header that appears after non-pseudo headers.");
+            }
+            if (pseudoHeaders[++i] == null) {
+                throw new IllegalArgumentException("pseudoHeaders value at index " + i + " is null");
+            }
+        }
+        for (i = 0; i < otherHeaders.length; ++i) {
+            name = otherHeaders[i];
+            HTTP2_NAME_VALIDATOR.validateName(name);
+            if (!seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) != ':') {
+                seenNonPseudoHeader = true;
+            } else if (seenNonPseudoHeader && !name.isEmpty() && name.byteAt(0) == ':') {
+                throw new IllegalArgumentException(
+                     "otherHeaders name at index " + i + " is a pseudo header that appears after non-pseudo headers.");
+            }
+            if (otherHeaders[++i] == null) {
+                throw new IllegalArgumentException("otherHeaders value at index " + i + " is null");
+            }
+        }
+    }
+
+    private AsciiString get0(CharSequence name) {
+        final int nameHash = name.hashCode();
+        AsciiString roName;
+        int roHash;
+        final int psudoHeadersEnd = pseudoHeaders.length - 1;
+        for (int i = 0; i < psudoHeadersEnd; i += 2) {
+            roName = pseudoHeaders[i];
+            roHash = roName.hashCode();
```

Comment:
yip this should be `AsciiString.hashCode()` ... This may incur additional costs if the hash code is not cached (e.g. `String`) so I may just pull this out...